### ISSUE TITLE
Assignment casts

### ISSUE THREAD

**Issue by vpetrovykh** [edgedb/edgedb#376]
Assignment casts are in a lot of ways a convenience and usability feature. For example, we're pretty sure that we want to allow casting a `str` into `uuid` or `datetime` on assignment (such as in `INSERT` or `UPDATE`). It just reduces the verbosity.

Things get a bit weirder with numerical scalars, there we also seem to want to be able to assign any numerical type to any other one (as long as the explicit cast exists). Let's say we failed to find an implicit cast that does the trick, so we start looking for an assignment cast (maybe from `float64` to `int64` like in `foo := 3^5`). Unlike the directed graph of implicit casts that had no cycles, the assignment cast graph necessarily has cycles (remember that technically any numeric is fair game as assignment to any other numeric).

Every hop on the implicit cast graph is technically "safe". It incurs a slight performance penalty, but the idea is that in principle the data does not degrade, nor does it cause exceptions. Assignment casts are not safe: they can degrade the data (`float64` -> `int64`) or even cause exceptions (`str` -> `int64`) on some inputs. So it seems that it is no longer a safe assumption that just because *A* is assignment-castable to *B* and *B* is assignment-castable to *C* that it's safe to conclude that *A* is assignment-castable to *C*. For example imagine that we decide one day that `uuid` is assignment-castable to `str`. This in no way implies that it is possible to assignment-cast `uuid` to `int64`. So assignment casts are *non-transitive*. This means that unless we have an explicit one-hop assignment cast declared between two types, it's not fair game to say there is one.

The above is complicated by the fact that we may be dealing with types that are derived from the basic scalars, so they will not have explicit assignment casts declared for the exact types. So we would need to look for possible valid casts specifically between the earliest non-abstract parent types.

**Comment by 1st1** [comment]
So what is this issue about? Do we have a bug to fix? What's the action?

**Comment by vpetrovykh** [comment]
Are we in agreement that assignment casts are non-transitive? I can adjust the algorithms for them and that would fix currently failing assignment cast tests. 

Basically the algorithm for finding assignment cast distance would be:
For every possible assignment cast add the implicit cast distance from the source to cast source and the distance from the cast target to target + 1 (the +1 is for the single assignment cast jump itself). From this point on the distance can be used in our existing assignment cast algorithms. This would allow defining cyclical assignment casts without causing ambiguity since only one hop would be used anyway.

**Comment by vpetrovykh** [comment]
Oh, yeah, the assignment cast distance is the minimum distance from the above.

**Comment by 1st1** [comment]
Can this be closed now?

**Comment by vpetrovykh** [comment]
After some more discussion we arrived at the decision that assignment casts are mainly needed to allow a few common cases of assigning a numeric literal to some property that has fewer bits that the literal default. This would not be an implicit cast, but it seems odd to require an explicit one, so that means that the only assignment casts are these:

- `int64` -> `int32`, `int16`, `float32`
- `float64` -> `float32`

For the other cases an explicit cast would still be needed if an implicit one would not do the trick. The assignment cast should be explicitly defined between the source and target base scalars without a need for implicit casts along the way. This simplifies the code and general logic of what can be expected in a given situation.

This has been implemented in the #377 PR.
